render/render.js

```diff
@@ -684,17 +684,17 @@ module.exports = function() {
 			if (key === "value") {
 				// Only do the coercion if we're actually going to check the value.
 				/* eslint-disable no-implicit-coercion */
-				var isFileInput = vnode.tag === "input" && vnode.attrs.type === "file"
 				//setting input[value] to same value by typing on focused element moves cursor to end in Chrome
 				//setting input[type=file][value] to same value causes an error to be generated if it's non-empty
-				if ((vnode.tag === "input" || vnode.tag === "textarea") && vnode.dom.value === "" + value && (isFileInput || vnode.dom === activeElement(vnode.dom))) return
+				//setting input[value] to same value sometimes causes checkValidity() to return true when form fields are invalid(#2256)
```

Comment:
The precise issue is this: `minlength` and `maxlength` constraints in form controls are only checked against values generated from user edits, not from script edits. When a user edits it and a script reads and writes the same previous value, the current value is considered to be last edited by script, and thus `minlength` and `maxlength` constraints are no longer checked.

I'll leave it to you how to distill this into a comment. Here's the relevant spec links:

- https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-maxlength
- https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-minlength